### DISCUSSION TITLE
History slider discussion

### DISCUSSION THREAD

**Discussion by tjomson** [git-truck/git-truck#729] [Research]
notes related to adding history slider to the truck

**Comment by tjomson** [comment]
Reasons not to use duckdb-wasm
-Not working offline in firefox (attempts to fetch some resource on db.instantiate)
-Seemingly very high memory utilization (mem leak?)
-Can't bundle wasm with remix (possible with vite, but remix-vite is unstable)
-Duckdb-wasm defaults to single-threaded execution (due to mitigation of spectre and meltdown)

**Comment by tjomson** [comment]
Query performance is quite good. Running the following query on the linux repo takes 103 ms.
```sql
SELECT author, COUNT(*) as commit_count
FROM commits
GROUP BY author
ORDER BY commit_count DESC
LIMIT 10;
```
Result:
![image](https://github.com/git-truck/git-truck/assets/23435481/0aa6c905-e96c-40ff-9d9f-83adb2e34f79)

The main bottleneck currently is data ingestion, probably due to having to construct the query string

**Reply by tjomson** [reply]
Noticing now that Mark Raasveldt is included. He should not be. This is probably because the database needs to be destroyed after use, or if multiple instances of GT are running, the db instances need to be separate

**Reply by mircealungu** [reply]
Who is Mark R? What's wrong with him?

**Comment by tjomson** [comment]
Ideas for necessary tables

renames
- from
- to
- date

filechanges
- contribs
- path
- author
- isbinary
- commit-hash (foreign key to commits table)
- timestamp ?

files
- creation-date
- deletion-date (nullable)
- path
- commits (foreign keys)

commits
- date
- hash
- message
- body
- timestamp
- list of foreign keys to filechanges table

author aliases
- alias (key)
- actual name

**Comment by tjomson** [comment]
I revised the data model a bit, as it became apparent that storing file sizes at given times and when a file is created/deleted is unpractical for the following reasons:
-git log can output either contributions (--numstat) or update status (--name-status) but not both at once.
-We need the file size at a given point, which only ls-tree can supply.

Thus the proposed data model is:

```sql
CREATE TABLE IF NOT EXISTS commits (
      hash VARCHAR PRIMARY KEY,
      author VARCHAR,
      time UINTEGER,
      body VARCHAR,
      message VARCHAR
    );
    CREATE TABLE IF NOT EXISTS filechanges (
      commithash VARCHAR,
      author VARCHAR, --Duplicate information, but might enable faster querying, as joining with commits-table is not necessary
      contribcount UINTEGER,
      filepath VARCHAR,
      isbinary BOOLEAN,
      FOREIGN KEY (commithash) REFERENCES commits(hash)
    );
    CREATE TABLE IF NOT EXISTS authorunions (
      alias VARCHAR PRIMARY KEY,
      actualname VARCHAR
    );
    CREATE TABLE IF NOT EXISTS renames (
      from VARCHAR,
      to VARCHAR,
      timestamp UINTEGER
    );
```

**Comment by tjomson** [comment]
Problem: How do we best handle co-authors?
My best idea is to add multiple entries to the filechanges table, and possibly add a boolean field "iscoauthor", such that we can filter duplicate entries out when counting the number of commits for a file

**Comment by tjomson** [comment]
For some reason, I need to increase the stack size of nodejs to avoid a crash. This can be done this the flag --stack-size=4096, but we can not decide which flags the user runs GT with.
I thought that a possible solution could be to use `setFlagsFromString('--stack-size=4096')` to set it programatically, but just like th documentation says, this may not work, which in my case it doesn't.
We need to find some other way to increase the stack size, or fix whatever it is that requires a large stack

**Comment by tjomson** [comment]
I thought that maybe doing Promise.all() to get information from the db would be faster due to parallelization in duckdb, but it seems like either the bottleneck is the single threaded nature of nodejs, or duckdb can only handle a single query at a time.
Which ever it is, Promise.all() performs the same as awaiting each function call individually

**Comment by tjomson** [comment]
When the first rename in a chain occurs (A->B) at time T, then my program will make an entry that says that any changes to A from time 0 to T-1 should go to B.
The problem with this is that if a file with name A had existed earlier and was deleted, then those credits would also go to B, which is unintentional.
Either we need to register whenever a file is created and deleted, or consider this an edge case not to worry about.

**Reply by mircealungu** [reply]
Can't we model the time a file was added to the repository? Ah. This is what you suggest in the last sentence. I think that would indeed solve the problem. Would it result in a too complicated algorithm? It would be nice if we had a data structure that was precise and fast at the same time!

I guess my idea with creating an object for every file and then adding the file names as temporally annotated labels to it would still need the same data: when was the file object created, and when was it deleted...

**Reply by tjomson** [reply]
>Can't we model the time a file was added to the repository? 

Yes, we can. I have to look into how to get git log to also show whether a file was created/modified/deleted.

**Reply by mircealungu** [reply]
It seems that we'll have to do that!

**Comment by tjomson** [comment]
When installing GT, pre-gyp will look for a duckdb binary that matches the system. If it exists, then that binary will be downloaded, without the system having to compile duckdb  itself. If not, then node-gyp will compile the source code for the system, however this has the downside of requiring python.
Luckily there are binaries for linux, windows, and macos, so python is therefore not needed for our target audience.
Using GT on android, will however not be possible any more.

**Reply by mircealungu** [reply]
I remember the pride with which one of you was showing me GT for Android :) 
That being said, I don't think there was anybody else besides you, and even you didn't use that platform too often I suspect :)

**Reply by tjomson** [reply]
Yes, Jonas found the bug when he tried running it on his phone :D

**Comment by tjomson** [comment]
I tried benchmarking my way of building a huge insert-query against the way the duckdb docs suggests, which is 
```js
con.run('CREATE TABLE a (i INTEGER)');
const stmt = con.prepare('INSERT INTO a VALUES (?)');
for (let i = 0; i < 10; i++) {
  stmt.run(i);
}
stmt.finalize();
```
turns out my way is a lot faster. About 1.5 sec vs 13 sec to insert 2.5k commits. Still quite slow though.

**Comment by tjomson** [comment]
I have made a new method for following renames, that is a lot nicer to explain, and is also a lot faster.
It has a number of steps, and it is apparent, that what now dominates the time, is inserting the renames into the database.

getcurrentrenames: 71.779ms
getfiles: 43.887ms
generatechains: 23.734ms
flatten: 15.538ms
replace: 1.117s

As you can see, the step "replace" which puts the rows into the database, takes 87% of the total time.
This is another reason why finding a faster way of inserting would be great.

**Reply by mircealungu** [reply]
> a lot nicer to explain, and is also a lot faster

I like this ^^^^ 

the times are, i suspect, for Git-Truck? 

do you have ideas of other ways? if you have a lot of rows, it will unfortunately take a long time :(

**Reply by tjomson** [reply]
I don't remember exactly which repo i tested on, it was either GT or duckdb. The main point is that database insertions take up a lot of time :/
I have tested out another method that is a lot faster, but that one has the downside of not working on Windows, because the Apache Arrow extension for duckdb apparently doesn't work on Windows.

**Comment by mircealungu** [comment]
Argh. Is there a roadmap and hope for windows? We could have an OS
dependent implementation?

M.

**Reply by tjomson** [reply]
Their documentation says "doesn't work on Windows **yet**", and they have a github action to build for windows, that is just not run on release. I guess that means, they are planning for it in the future.
Otherwise we can check the system, and use the fast method on mac/linux and the slow on windows.
It will be just like in the old days, when GT was much slower on windows than mac/linux.

**Reply by tjomson** [reply]
Good news!
I just made another method that is roughly as fast a the way that isn't supported on Windows.
It works by writing json to the disk, and then ingesting it into duckdb, so the reason that it is fast might be that my pc has a fast ssd.
I got some benchmarking to do 😎

**Comment by mircealungu** [comment]
Aiii iaaa. Wasn’t writing very big JSON files the reason for which we didn’t scale on large systems?


Sent with a Spark

**Reply by tjomson** [reply]
Indeed it was. Luckily, this time we do not need to write huge individual json-files. I can split them up into appropriate sizes.
It feels quite silly to write something from memory to disk, just to read it into memory again and delete the file, but if it works, it works.
Avoiding doing that, would be by using apache arrow, which is currently unsupported on windows.

**Comment by mircealungu** [comment]
I'd it works for Linux then it really works :)

**Comment by tjomson** [comment]
I should have done this a long time ago, but I have finally checked whether duckdb is actually faster than sqlite. Luckily for us, it is.
Results for tensorflow with full time range:

|                       | duckdb | sqlite |
|-----------------------|--------|--------|
| Generate big cache    | 967 ms | 4.64 s |
| Dominant author query | 213 ms | 813 ms |
| Commit count query    | 171 ms | 552 ms |

**Comment by mircealungu** [comment]
Nice :) better later than never !!!
